dali/pipeline/op_graph.h

```diff
@@ -298,33 +279,85 @@ class DLL_PUBLIC OpGraph {
    * @brief Save graph in DOT directed graph format
    * in filename.
    */
-  DLL_PUBLIC void SaveToDotFile(const string filename) {
+  DLL_PUBLIC void SaveToDotFile(const string filename, bool show_tensors = false,
+                                bool show_ids = false, bool use_colors = false) {
     std::ofstream ofs(filename);
     ofs << "digraph graphname {\n";
-    const OpNode& current_node = GetNodeForIdx(0);
-    GenerateDOTFromGraph(current_node, ofs);
+    GenerateDOTFromGraph(ofs, show_tensors, show_ids, use_colors);
     ofs << "}\n";
-    visited_nodes_.clear();
   }
 
   DISABLE_COPY_MOVE_ASSIGN(OpGraph);
 
  private:
-  vector<OpNode> cpu_nodes_;
-  vector<OpNode> gpu_nodes_;
-  vector<OpNode> mixed_nodes_;
-  vector<OpNode> support_nodes_;
+  // Should be called only once for each tensor
+  void GenerateDOTFromGraph(const TensorNode& current_node, std::ofstream& ofs, bool show_tensors,
+                            bool show_ids);
+
+  /**
+   * @brief Recalculate OpNodes partitioning
+   *
+   * Clears the partition vectors and readds all the nodes to proper partitions again,
+   * storing new indexes.
+   */
+  void Repartition();
 
-  // Stores a mapping from NodeIDs to a pair where the first
-  // element indicates what type of node it is,  and the second
-  // is the index of the op within the specified vector.
-  vector<std::pair<OpType, Index>> id_to_node_map_;
+  /**
+   * @brief Adds new OpNode of `op_type` to op_nodes_ unified vector and to proper partition
+   * storing its spec and instance_name. OpNode are given consecutive ids.
+   *
+   * @return OpNode& Reference to added OpNode.
+   */
+  OpNode& PlaceNewOp(OpType op_type, OpSpec op_spec, std::string instance_name);
 
-  std::map<string, TensorMeta> tensor_producers_;
-  std::map<string, vector<TensorMeta>> tensor_consumers_;
+  /**
+   * @brief Creates new tensor node with conscutive id.
+   *
+   * @return TensorNode& Reference to added tensor node.
```

Comment:
```suggestion
   * @return TensorNode& Reference to the newly added tensor node.
```
BTW - does retyping the type here do any good?

Reply:
Done.
I didn't retype it, my extension for IDE did it. Removed, it does not help in any way.